### ISSUE TITLE
Repetition of code when multiple modules are using the same path.

### ISSUE THREAD

**Issue by ekouts** [reframe-hpc/reframe#2604]
I am opening this after a conversation with @jgphpc :
When we define module with the full path there can be unnecessary duplication in the code:
`config.py`
```python
...
            'modules': [
                {
                    'name': 'cray-mpich-binary/8.1.18.4-gcc-twtq6ct',
                    'path': '/scratch/e1000/gcc/11.3.0',
                },
                {
                    'name': 'cuda/11.7.1-yg5gvzu.lua',
                    'path': '/scratch/e1000/gcc/11.3.0',
                },
             ]
...
```

ReFrame will generate this in the script:
```bash
module use /scratch/e1000/gcc/11.3.0
module load cray-mpich-binary/...
module use /scratch/e1000/gcc/11.3.0
module load cuda/...
```

If we remove the path from the second one ReFrame will complain
```python
...
            'modules': [
                {
                    'name': 'cray-mpich-binary/8.1.18.4-gcc-twtq6ct',
                    'path': '/scratch/e1000/gcc/11.3.0',
                },
                {
                    'name': 'cuda/11.7.1-yg5gvzu.lua',
                    # 'path': '/scratch/e1000/gcc/11.3.0',
                },
             ]
...
```
with this
```
Lmod Warning: Failed to find the following module(s): "cuda/11.7.1-yg5gvzu" in your MODULEPATH
```

Maybe we could do something better in reframe so that it doesn't try to add the module path unnecessarily.

**Comment by jgphpc** [comment]
Likewise, it would be nice to be able to mix definitions:
```
...
 'modules': [
                {
                    'name': 'cray-mpich-binary/8.1.18.4-gcc-twtq6ct',
                    'path': '/scratch/e1000/gcc/11.3.0',
                },
                'cuda/11.7.1-yg5gvzu.lua',
             ]
...
```

Reframe will refuse:

**Comment by jgphpc** [comment]
@ekouts suggests:

**Comment by RiMillo** [comment]
I agree with the enhancements proposed above. I would also suggest to have multiple paths (I dealing with packages installed with spack which used an upstream, so both the modulepath of the spack environment and the upstream should be should be added):
```
{ "name": "foo/1.0", "paths": ["/path/to/spack/upstream", "/path/to/spack/env"] }
```

To be honest, I was able to make it work just by using merging them:
```
{ "name": "foo/1.0", "path": "/path/to/spack/upstream /path/to/spack/env" } # Notice blank between the two paths
```

In any case, it was quite tricky to find that it is possible to use `path` with module definition in the environment / system configuration. I did not find it explicitly mentioned in the doc (see [here](https://reframe-hpc.readthedocs.io/en/stable/howto.html#working-with-environment-modules) or [here](https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#reframe.core.environments.Environment.modules)) and had to dig into the source. It might be worth it to improve the doc a bit.

**Comment by vkarak** [comment]
Ok, I see 3-4 issues already mentioned here :-) We should rather split in multiple ones and link to this one:

1. The original issue reported by @ekouts 
2. The extra enhancement requested by @jgphpc (https://github.com/reframe-hpc/reframe/issues/2604#issuecomment-1257907821)
3. The additional enhancement requested by @RiMillo (https://github.com/reframe-hpc/reframe/issues/2604#issuecomment-5343592802)
4. The documentation improvement raised by @RiMillo.
